### PR TITLE
MODORDERS-39 - Implement GET /orders/id

### PR DESCRIPTION
**Pull request by craigmcnally**
Implement the GET /orders/id endpoint.  See [MODORDERS-39](https://issues.folio.org/browse/MODORDERS-39) and [MODORDERS-41](https://issues.folio.org/browse/MODORDERS-41)

Note:  this work also fixes [MODORDERS-54](https://issues.folio.org/browse/MODORDERS-54)

### CODE REVIEW

**Comment by craigmcnally** [comment]
This should bring our test coverage back above 80%.  I would like to spend some time improving the unit test assertions, but don't want to hold up getting this in.  If I have time I'll get some of that into this PR.  If not I'll create a follow-up user story

File: src/main/java/org/folio/rest/impl/GetOrdersByIdHelper.java [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+import io.vertx.core.http.HttpMethod;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import me.escoffier.vertx.completablefuture.VertxCompletableFuture;
+
+public class GetOrdersByIdHelper {
+
+  private static final Logger logger = Logger.getLogger(GetOrdersByIdHelper.class);
+
+  public static final String BASE_MOCK_DATA_PATH = "mockdata/";
```
Comment (KVupp) [inline src/main/java/org/folio/rest/impl/GetOrdersByIdHelper.java]: Are we mocking data in implementation class?

Comment (KVupp) [inline src/main/java/org/folio/rest/impl/GetOrdersByIdHelper.java]: This variable needs to be removed?

File: src/main/java/org/folio/rest/impl/GetOrdersByIdHelper.java [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    this.ctx = ctx;
+    this.asyncResultHandler = asyncResultHandler;
+  }
+
+  public CompletableFuture<CompositePurchaseOrder> getOrder(String id, String lang) {
+    CompletableFuture<CompositePurchaseOrder> future = new VertxCompletableFuture<>(ctx);
+
+    getCompositePurchaseOrderById(id, lang)
+      .thenAccept(orders -> {
+        logger.info("Returning mock data: " + JsonObject.mapFrom(orders).encodePrettily());
```
Comment (KVupp) [inline src/main/java/org/folio/rest/impl/GetOrdersByIdHelper.java]: The info message must be changed

File: src/main/java/org/folio/rest/impl/GetOrdersByIdHelper.java [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    return future;
+  }
+
+  private CompletableFuture<CompositePurchaseOrder> getCompositePurchaseOrderById(String id, String lang) {
+    CompletableFuture<CompositePurchaseOrder> future = new VertxCompletableFuture<>(ctx);
+
+    getPurchaseOrder(id, lang)
+      .thenAccept(po -> {
+        logger.info("got: " + po.encodePrettily());
+        po.remove("adjustment");
```
Comment (KVupp) [inline src/main/java/org/folio/rest/impl/GetOrdersByIdHelper.java]: Isn't the adjustment required to be returned for this API?

File: src/main/java/org/folio/rest/impl/GetOrdersByIdHelper.java (hunk 2 above)
Comment (craigmcnally) [reply src/main/java/org/folio/rest/impl/GetOrdersByIdHelper.java]: good call

File: src/main/java/org/folio/rest/impl/GetOrdersByIdHelper.java [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+  public CompletableFuture<CompositePurchaseOrder> getOrder(String id, String lang) {
+    CompletableFuture<CompositePurchaseOrder> future = new VertxCompletableFuture<>(ctx);
+
+    getCompositePurchaseOrderById(id, lang)
+      .thenAccept(orders -> {
+        logger.info("Returning mock data: " + JsonObject.mapFrom(orders).encodePrettily());
+        future.complete(orders);
+      })
+      .exceptionally(t -> {
+        logger.error("Error getting orders", t);
```
Comment (KVupp) [inline src/main/java/org/folio/rest/impl/GetOrdersByIdHelper.java]: probably trivial  but wouldn't it be easier to debug if we specify the error message that  we are trying to get order by id

File: src/main/java/org/folio/rest/impl/GetOrdersByIdHelper.java (hunk 3 above)
Comment (KVupp) [inline src/main/java/org/folio/rest/impl/GetOrdersByIdHelper.java]: don't we need the adjustment at the PO level?

File: src/main/java/org/folio/rest/impl/GetOrdersByIdHelper.java (hunk 4 above)
Comment (craigmcnally) [reply src/main/java/org/folio/rest/impl/GetOrdersByIdHelper.java]: You get that for free if using the default log4j.properties:

`16:29:56 ERROR GetOrdersByIdHelper  Error getting orders
java.util.concurrent.CompletionException: org.folio.orders.rest.exceptions.HttpException: non-existent-po-id`

File: src/main/java/org/folio/orders/utils/HelperUtils.java [hunk 5]
```diff
@@ -23,7 +31,47 @@ public static String getMockData(String path) throws IOException {
...
+
+    return response.getBody();
+  }
+
+  public static Adjustment calculateAdjustment(List<PoLine> lines) {
+    Adjustment ret = new Adjustment();
+    for (int i = 0; i < lines.size(); i++) {
+      Adjustment a = lines.get(i).getAdjustment();
+      if (a != null) {
+        if (ret == null) {
```
[the remark this answers is not in the document]
Comment (craigmcnally) [reply src/main/java/org/folio/orders/utils/HelperUtils.java]: another good catch.

File: src/main/java/org/folio/rest/impl/GetOrdersByIdHelper.java [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+
+  private CompletableFuture<PoLine> resolvePoLine(JsonObject line) {
+    CompletableFuture<PoLine> future = new VertxCompletableFuture<>(ctx);
+
+    List<CompletableFuture<Void>> futures = new ArrayList<>();
+    futures.add(resolveSubObjIfPresent(line, "adjustment", "/adjustment/"));
+    futures.add(resolveSubObjIfPresent(line, "cost", "/cost/"));
+    futures.add(resolveSubObjIfPresent(line, "details", "/details/"));
+    futures.add(resolveSubObjIfPresent(line, "eresource", "/eresource/"));
+    futures.add(resolveSubObjIfPresent(line, "location", "/location/"));
```
Comment (KVupp) [inline src/main/java/org/folio/rest/impl/GetOrdersByIdHelper.java]: Isn't this URL /locations? https://github.com/folio-org/mod-inventory-storage/blob/master/ramls/location.raml , or Am I looking the wrong API?

Comment (craigmcnally) [reply src/main/java/org/folio/rest/impl/GetOrdersByIdHelper.java]: nope I think you're correct.  I had a feeling I was going to get burned by the competing plural/singular api conventions eventually

**Pull request merged**